File: packages/openmemory-py/src/openmemory/utils/text.py
import re
import hashlib
from typing import List, Set, Dict

from ..core.config import env
from ..i18n import get_lang_pack


_LANG_PACK = get_lang_pack(env.lang, env.domain)

CMAP: Dict[str, str] = {}
SLOOK: Dict[str, Set[str]] = {}

for grp in _LANG_PACK.synonym_groups:
    can = grp[0]
    sset = set(grp)
    for w in grp:
        CMAP[w] = can
        SLOOK[can] = sset


CJK_PAT = r"㐀-䶿一-鿿豈-﫿぀-ヿ가-힯"
TOK_PAT = re.compile(rf"[a-z0-9]+|[{CJK_PAT}]+", re.I)


def _expand_cjk_token(tok: str) -> List[str]:
    if len(tok) <= 1:
        return [tok]
    return [tok[i : i + 2] for i in range(len(tok) - 1)]


def tokenize(text: str) -> List[str]:
    res: List[str] = []
    for tok in TOK_PAT.findall(text):
        low = tok.lower()
        if re.fullmatch(rf"[{CJK_PAT}]+", tok):
            res.extend(_expand_cjk_token(tok))
        else:
            res.append(low)
    return res


def stem(tok: str) -> str:
    if not _LANG_PACK.should_stem(tok):
        return tok
    return _LANG_PACK.stem(tok)


def canonicalize_token(tok: str) -> str:
    if not tok:
        return ""
    low = tok.lower()
    if low in CMAP:
        return CMAP[low]
    st = stem(low)
    return CMAP.get(st, st)


def canonical_tokens_from_text(text: str) -> List[str]:
    res = []
    for tok in tokenize(text):
        can = canonicalize_token(tok)
        if can and len(can) > 1:
            res.append(can)
    return res


def synonyms_for(tok: str) -> Set[str]:
    can = canonicalize_token(tok)
    return SLOOK.get(can, {can})


def build_search_doc(text: str) -> str:
    can = canonical_tokens_from_text(text)
    exp = set()
    for tok in can:
        exp.add(tok)
        syns = SLOOK.get(tok)
        if syns:
            exp.update(syns)
    return " ".join(exp)


def build_fts_query(text: str) -> str:
    can = canonical_tokens_from_text(text)
    if not can:
        return ""
    uniq = sorted(list(set(t for t in can if len(t) > 1)))
    return " OR ".join(f'"{t}"' for t in uniq)


def canonical_token_set(text: str) -> Set[str]:
    return set(canonical_tokens_from_text(text))


def stable_text_fallback_hash(text: str) -> str:
    return hashlib.blake2b(text.encode("utf-8"), digest_size=8).hexdigest()

File: packages/openmemory-js/src/utils/text.ts
import crypto from "node:crypto";
import { env } from "../core/cfg";
import { getLangPack } from "../i18n";

const lang_pack = getLangPack(env.lang, env.domain);

const cmap = new Map<string, string>();
const slook = new Map<string, Set<string>>();

for (const grp of lang_pack.synonymGroups) {
    const can = grp[0];
    const sset = new Set(grp);
    for (const w of grp) {
        cmap.set(w, can);
        slook.set(can, sset);
    }
}

const cjk_pat =
    /[㐀-䶿一-鿿豈-﫿぀-ヿ가-힯]+/u;
const tok_pat =
    /[a-z0-9]+|[㐀-䶿一-鿿豈-﫿぀-ヿ가-힯]+/giu;

const expand_cjk_token = (tok: string): string[] => {
    if (tok.length <= 1) return [tok];
    const expanded: string[] = [];
    for (let i = 0; i < tok.length - 1; i++) {
        expanded.push(tok.slice(i, i + 2));
    }
    return expanded;
};

export const tokenize = (text: string): string[] => {
    const toks: string[] = [];
    let m: RegExpExecArray | null;
    while ((m = tok_pat.exec(text))) {
        const tok = m[0];
        if (cjk_pat.test(tok)) {
            toks.push(...expand_cjk_token(tok));
            continue;
        }
        toks.push(tok.toLowerCase());
    }
    return toks;
};

const stem = (tok: string): string => {
    if (!lang_pack.shouldStem(tok)) return tok;
    return lang_pack.stem(tok);
};

export const canonicalize_token = (tok: string): string => {
    if (!tok) return "";
    const low = tok.toLowerCase();
    if (cmap.has(low)) return cmap.get(low)!;
    const st = stem(low);
    return cmap.get(st) || st;
};

export const canonical_tokens_from_text = (text: string): string[] => {
    const res: string[] = [];
    for (const tok of tokenize(text)) {
        const can = canonicalize_token(tok);
        if (can && can.length > 1) {
            res.push(can);
        }
    }
    return res;
};

export const synonyms_for = (tok: string): Set<string> => {
    const can = canonicalize_token(tok);
    return slook.get(can) || new Set([can]);
};

export const build_search_doc = (text: string): string => {
    const can = canonical_tokens_from_text(text);
    const exp = new Set<string>();
    for (const tok of can) {
        exp.add(tok);
        const syns = slook.get(tok);
        if (syns) {
            syns.forEach((s) => exp.add(s));
        }
    }
    return Array.from(exp).join(" ");
};

export const build_fts_query = (text: string): string => {
    const can = canonical_tokens_from_text(text);
    if (!can.length) return "";
    const uniq = Array.from(new Set(can.filter((t) => t.length > 1)));
    return uniq.map((t) => `"${t}"`).join(" OR ");
};

export const canonical_token_set = (text: string): Set<string> => {
    return new Set(canonical_tokens_from_text(text));
};

export const stable_text_fallback_hash = (text: string): string => {
    return crypto
        .createHash("blake2b512")
        .update(text, "utf8")
        .digest("hex")
        .slice(0, 16);
};

export const add_synonym_tokens = (toks: Iterable<string>): Set<string> => {
    const res = new Set<string>();
    for (const tok of toks) {
        res.add(tok);
        const syns = slook.get(tok);
        if (syns) {
            syns.forEach((s) => res.add(canonicalize_token(s)));
        }
    }
    return res;
};
